### PR TITLE
[IMP] account_payment,sale: don't assign all SO's tx on invoice creation

### PR DESCRIPTION
**Pull request by Demesmaeker**
task-3706978

### CODE REVIEW

**Review by vchu-odoo — APPROVED** [review]
LGTM :+1: 

Do you think that adding test could be useful or not?

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -1152,6 +1152,11 @@ def _prepare_invoice(self):
         """
         self.ensure_one()
 
+        txs_to_be_linked = self.transaction_ids.filtered(
+            lambda tx: (tx.state == 'done' and not (tx.payment_id and tx.payment_id.is_reconciled))
+                       or tx.state in ('pending', 'authorized')
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:1157]:
```suggestion
            lambda tx: (
                tx.state in ('pending', 'authorized')
                or tx.state == 'done'
                and not (tx.payment_id and tx.payment_id.is_reconciled)
            )
```
feels clearer to me, but feel free to find another clean way to display it (I feel like the display is not really clear rn)

File: addons/sale/tests/test_payment_flow.py [hunk 2]
```diff
@@ -300,6 +300,70 @@ def test_invoice_is_final(self):
...
+        linked_txs = first_invoice.transaction_ids
+        msg = "The newly created invoice should be linked to the done and pending transactions."
+        self.assertEqual(len(linked_txs), 2, msg=msg)
+        self.assertTrue(linked_txs in (partial_tx_done, partial_tx_pending), msg=msg)
+        msg = "The payment is not yet reconciled."
+        self.assertFalse(partial_tx_done.payment_id.is_reconciled, msg=msg)
+
+        partial_tx_done._reconcile_after_done()
+
+        msg = "The payment is now reconciled."
```
Comment (Feyensv) [inline addons/sale/tests/test_payment_flow.py:356]: The message is displayed when the assertion fails, not when the assertion succeeds :eyes:

File: addons/sale/tests/test_payment_flow.py [hunk 3]
```diff
@@ -300,6 +300,70 @@ def test_invoice_is_final(self):
...
+        self.sale_order._create_invoices()
+
+        self.assertEqual(len(self.sale_order.invoice_ids), 1, msg="1 invoice is created.")
+
+        first_invoice = self.sale_order.invoice_ids
+        linked_txs = first_invoice.transaction_ids
+        msg = "The newly created invoice should be linked to the done and pending transactions."
+        self.assertEqual(len(linked_txs), 2, msg=msg)
+        self.assertTrue(linked_txs in (partial_tx_done, partial_tx_pending), msg=msg)
+        msg = "The payment is not yet reconciled."
```
Comment (Feyensv) [inline addons/sale/tests/test_payment_flow.py:351]: same

File: addons/sale/tests/test_payment_flow.py [hunk 4]
```diff
@@ -300,6 +300,70 @@ def test_invoice_is_final(self):
...
+            partial_tx_done._reconcile_after_done()
+        partial_tx_pending = self._create_transaction(
+            flow='direct',
+            amount=2,
+            sale_order_ids=[self.sale_order.id],
+            state='pending',
+            reference='partial_tx_pending',
+        )
+        self.assertTrue(partial_tx_done.payment_id, msg="Account payment should have been created.")
+        msg = "The created account payment is not yet reconciled as there are no invoice yet."
```
Comment (Feyensv) [inline addons/sale/tests/test_payment_flow.py:324]: same

File: addons/sale/models/sale_order.py (hunk 1 above)
Comment (Demesmaeker) [reply addons/sale/models/sale_order.py:1157]: I understand why you think so, but I don't like getting rid of the parenthesis like that, even if correct, for clarity purpose too. Would this be okay for you? As they are still on the same line, even without parenthesis, it's still clear it's the same condition
```suggestion
            lambda tx: (
                tx.state in ('pending', 'authorized')
                or tx.state == 'done' and not (tx.payment_id and tx.payment_id.is_reconciled)
            )
```

**Comment by Feyensv** [comment]
@robodoo rebase-ff

**Pull request closed without merging**